File: runtimes/python/src/wrapper_runtime/auth/oauth_loopback.py
"""T054: OAuth 2.1 + PKCE loopback flow for the Python runtime (FR-014, FR-016).

Mirrors runtimes/node/src/auth/oauth-loopback.ts. Built on `authlib` in production; the
PKCE/URL machinery here is stdlib so it is testable without a live IdP.
"""
from __future__ import annotations

import base64
import hashlib
import http.server
import os
import secrets
import threading
import time
import urllib.parse
import urllib.request
from dataclasses import dataclass
from typing import Callable


@dataclass
class TokenSet:
    access_token: str
    refresh_token: str | None
    expires_at: float


def generate_pkce() -> tuple[str, str]:
    verifier = base64.urlsafe_b64encode(os.urandom(32)).rstrip(b"=").decode()
    challenge = (
        base64.urlsafe_b64encode(hashlib.sha256(verifier.encode()).digest()).rstrip(b"=").decode()
    )
    return verifier, challenge


def build_authorization_url(config: dict, redirect_uri: str, challenge: str, state: str) -> str:
    params = {
        "response_type": "code",
        "client_id": config["clientId"],
        "redirect_uri": redirect_uri,
        "scope": " ".join(config.get("scopes", [])),
        "code_challenge": challenge,
        "code_challenge_method": "S256",
        "state": state,
    }
    return config["authorizationEndpoint"] + "?" + urllib.parse.urlencode(params)


def _default_exchange(config: dict) -> Callable[[str, str, str], TokenSet]:
    def exchange(code: str, verifier: str, redirect_uri: str) -> TokenSet:
        data = urllib.parse.urlencode(
            {
                "grant_type": "authorization_code",
                "code": code,
                "redirect_uri": redirect_uri,
                "client_id": config["clientId"],
                "code_verifier": verifier,
            }
        ).encode()
        req = urllib.request.Request(config["tokenEndpoint"], data=data, method="POST")
        with urllib.request.urlopen(req) as resp:  # noqa: S310
            import json

            j = json.loads(resp.read().decode())
        return TokenSet(
            access_token=j["access_token"],
            refresh_token=j.get("refresh_token"),
            expires_at=time.time() + j.get("expires_in", 3600),
        )

    return exchange


def run_loopback_flow(
    config: dict,
    *,
    open_browser: Callable[[str], None] | None = None,
    exchange_code: Callable[[str, str, str], TokenSet] | None = None,
    port: int = 0,
) -> TokenSet:
    verifier, challenge = generate_pkce()
    state = secrets.token_hex(8)
    exchange = exchange_code or _default_exchange(config)
    result: dict[str, TokenSet] = {}

    class Handler(http.server.BaseHTTPRequestHandler):
        def do_GET(self) -> None:  # noqa: N802
            qs = urllib.parse.urlparse(self.path).query
            params = urllib.parse.parse_qs(qs)
            code = params.get("code", [None])[0]
            returned_state = params.get("state", [None])[0]
            if not code or returned_state != state:
                self.send_response(400)
                self.end_headers()
                return
            self.send_response(200)
            self.end_headers()
            self.wfile.write(b"<h1>Authorized. You may close this window.</h1>")
            redirect_uri = f"http://127.0.0.1:{self.server.server_address[1]}/callback"
            result["tokens"] = exchange(code, verifier, redirect_uri)

        def log_message(self, *_args: object) -> None:  # silence
            pass

    server = http.server.HTTPServer(("127.0.0.1", port), Handler)
    bound_port = server.server_address[1]
    redirect_uri = f"http://127.0.0.1:{bound_port}/callback"
    auth_url = build_authorization_url(config, redirect_uri, challenge, state)
    (open_browser or (lambda u: print(f"Open this URL to authorize:\n{u}")))(auth_url)

    thread = threading.Thread(target=server.handle_request)
    thread.start()
    thread.join(timeout=120)
    server.server_close()
    return result["tokens"]

File: runtimes/java/src/main/java/wrapper/auth/OAuthLoopback.java
package wrapper.auth;

import com.sun.net.httpserver.HttpServer;
import java.net.InetSocketAddress;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.util.Base64;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

/**
 * T055: OAuth 2.1 + PKCE loopback flow for the Java runtime (FR-014, FR-016). Mirrors the
 * Node/Python loopback flow. Built on Nimbus oauth2-oidc-sdk in production; the PKCE/URL
 * machinery here is JDK-only so it is testable without a live IdP.
 */
public final class OAuthLoopback {
  private static final SecureRandom RNG = new SecureRandom();

  private OAuthLoopback() {}

  public static final class TokenSet {
    public final String accessToken;
    public final String refreshToken;
    public final long expiresAt;

    public TokenSet(String accessToken, String refreshToken, long expiresAt) {
      this.accessToken = accessToken;
      this.refreshToken = refreshToken;
      this.expiresAt = expiresAt;
    }
  }

  public static String[] generatePkce() throws Exception {
    byte[] v = new byte[32];
    RNG.nextBytes(v);
    String verifier = Base64.getUrlEncoder().withoutPadding().encodeToString(v);
    byte[] digest = MessageDigest.getInstance("SHA-256").digest(verifier.getBytes(StandardCharsets.UTF_8));
    String challenge = Base64.getUrlEncoder().withoutPadding().encodeToString(digest);
    return new String[] {verifier, challenge};
  }

  public static String buildAuthorizationUrl(
      Map<String, Object> config, String redirectUri, String challenge, String state) {
    StringBuilder sb = new StringBuilder(String.valueOf(config.get("authorizationEndpoint")));
    sb.append("?response_type=code");
    sb.append("&client_id=").append(enc(String.valueOf(config.get("clientId"))));
    sb.append("&redirect_uri=").append(enc(redirectUri));
    sb.append("&code_challenge=").append(enc(challenge));
    sb.append("&code_challenge_method=S256");
    sb.append("&state=").append(enc(state));
    return sb.toString();
  }

  private static String enc(String s) {
    return URLEncoder.encode(s, StandardCharsets.UTF_8);
  }

  /** Run the loopback flow. tokenExchanger allows injecting a fake exchange in tests. */
  public static TokenSet runLoopbackFlow(Map<String, Object> config, TokenExchanger exchanger)
      throws Exception {
    String[] pkce = generatePkce();
    String verifier = pkce[0];
    String challenge = pkce[1];
    String state = Long.toHexString(RNG.nextLong());

    HttpServer server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
    int port = server.getAddress().getPort();
    String redirectUri = "http://127.0.0.1:" + port + "/callback";
    CompletableFuture<TokenSet> future = new CompletableFuture<>();

    server.createContext(
        "/callback",
        exchange -> {
          Map<String, String> params = queryParams(exchange.getRequestURI());
          String code = params.get("code");
          String returnedState = params.get("state");
          byte[] resp = "<h1>Authorized. You may close this window.</h1>".getBytes(StandardCharsets.UTF_8);
          exchange.sendResponseHeaders(code != null && state.equals(returnedState) ? 200 : 400, resp.length);
          exchange.getResponseBody().write(resp);
          exchange.close();
          if (code != null && state.equals(returnedState)) {
            try {
              future.complete(exchanger.exchange(code, verifier, redirectUri));
            } catch (Exception e) {
              future.completeExceptionally(e);
            }
          }
        });
    server.start();

    String authUrl = buildAuthorizationUrl(config, redirectUri, challenge, state);
    System.err.println("Open this URL to authorize:\n" + authUrl);

    try {
      return future.get();
    } finally {
      server.stop(0);
    }
  }

  private static Map<String, String> queryParams(URI uri) {
    Map<String, String> out = new java.util.HashMap<>();
    String q = uri.getQuery();
    if (q != null) {
      for (String pair : q.split("&")) {
        String[] kv = pair.split("=", 2);
        out.put(kv[0], kv.length > 1 ? java.net.URLDecoder.decode(kv[1], StandardCharsets.UTF_8) : "");
      }
    }
    return out;
  }

  /** Default token exchange over HTTP (production path). */
  public static TokenExchanger httpExchanger(Map<String, Object> config, HttpClient client) {
    return (code, verifier, redirectUri) -> {
      String body =
          "grant_type=authorization_code&code=" + enc(code) + "&redirect_uri=" + enc(redirectUri)
              + "&client_id=" + enc(String.valueOf(config.get("clientId"))) + "&code_verifier=" + enc(verifier);
      HttpRequest req =
          HttpRequest.newBuilder()
              .uri(URI.create(String.valueOf(config.get("tokenEndpoint"))))
              .header("content-type", "application/x-www-form-urlencoded")
              .POST(HttpRequest.BodyPublishers.ofString(body))
              .build();
      HttpResponse<String> resp = client.send(req, HttpResponse.BodyHandlers.ofString());
      com.google.gson.JsonObject j =
          new com.google.gson.Gson().fromJson(resp.body(), com.google.gson.JsonObject.class);
      long expiresIn = j.has("expires_in") ? j.get("expires_in").getAsLong() : 3600L;
      return new TokenSet(
          j.get("access_token").getAsString(),
          j.has("refresh_token") ? j.get("refresh_token").getAsString() : null,
          System.currentTimeMillis() + expiresIn * 1000L);
    };
  }

  public interface TokenExchanger {
    TokenSet exchange(String code, String verifier, String redirectUri) throws Exception;
  }
}
